src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -24,53 +26,83 @@
-    private readonly RazorLSPOptionsMonitor _razorLSPOptionsMonitor;
-    private readonly IClientConnection _clientConnection;
-    private readonly ILogger _logger;
+    private readonly IRazorDocumentMappingService _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+    private readonly ILogger _logger = loggerFactory.CreateLogger<RazorSemanticTokensInfoService>();
+    private readonly ITelemetryReporter? _telemetryReporter = telemetryReporter;
 
-    public RazorSemanticTokensInfoService(
+    private RazorSemanticTokensLegend? _razorSemanticTokensLegend;
+
+    public void ApplyCapabilities(VSInternalServerCapabilities serverCapabilities, VSInternalClientCapabilities clientCapabilities)
+    {
+        _razorSemanticTokensLegend = new RazorSemanticTokensLegend(clientCapabilities);
+
+        serverCapabilities.SemanticTokensOptions = new SemanticTokensOptions
+        {
+            Full = false,
+            Legend = _razorSemanticTokensLegend.Legend,
+            Range = true,
+        };
+    }
+
+    public async Task<SemanticTokens?> GetSemanticTokensAsync(
         IClientConnection clientConnection,
-        IRazorDocumentMappingService documentMappingService,
-        RazorLSPOptionsMonitor razorLSPOptionsMonitor,
-        LanguageServerFeatureOptions languageServerFeatureOptions,
-        IRazorLoggerFactory loggerFactory)
+        TextDocumentIdentifier textDocumentIdentifier,
+        Range range,
+        VersionedDocumentContext documentContext,
+        bool colorBackground,
+        CancellationToken cancellationToken)
     {
-        _clientConnection = clientConnection ?? throw new ArgumentNullException(nameof(clientConnection));
-        _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
-        _razorLSPOptionsMonitor = razorLSPOptionsMonitor ?? throw new ArgumentNullException(nameof(razorLSPOptionsMonitor));
-        _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+        _razorSemanticTokensLegend.AssumeNotNull();
 
-        if (loggerFactory is null)
+        var correlationId = Guid.NewGuid();
+        using var _ = _telemetryReporter?.TrackLspRequest(Methods.TextDocumentSemanticTokensRangeName, LanguageServerConstants.RazorLanguageServerName, correlationId);
+
+        var semanticTokens = await GetSemanticTokensAsync(clientConnection, textDocumentIdentifier, range, documentContext, correlationId, colorBackground, cancellationToken).ConfigureAwait(false);
+
+        var amount = semanticTokens is null ? "no" : (semanticTokens.Data.Length / 5).ToString(Thread.CurrentThread.CurrentCulture);
```

Comment:
>5 [](http://example.com/codeflow?start=83&length=1)

Should this be the "TokenSize" const already defined in this class? Same with two usages of magic number "5" below :)

Reply:
Fair call, I just copied this as is. Will update

Reply:
Yep, I saw it first in the original code location and was like "what's the magic number 5 and where does it come from"? :) The constant in the target class made sense.